src/compiler/binder.ts

```diff
@@ -1751,4 +1791,440 @@ namespace ts {
+                if ((<HeritageClause>node).token === SyntaxKind.ExtendsKeyword) {
+                    // An `extends` HeritageClause is ES6 syntax.
+                    transformFlags |= TransformFlags.AssertES6;
+                }
+                else {
+                    // An `implements` HeritageClause is TypeScript syntax.
+                    Debug.assert((<HeritageClause>node).token === SyntaxKind.ImplementsKeyword);
+                    transformFlags |= TransformFlags.AssertTypeScript;
+                }
+
+                break;
+
+            case SyntaxKind.ExpressionWithTypeArguments:
+                // An ExpressionWithTypeArguments is ES6 syntax, as it is used in the
+                // extends clause of a class.
+                transformFlags |= TransformFlags.AssertES6;
+
+                // If an ExpressionWithTypeArguments contains type arguments, then it
+                // is TypeScript syntax.
+                if ((<ExpressionWithTypeArguments>node).typeArguments) {
+                    transformFlags |= TransformFlags.AssertTypeScript;
+                }
+
+                break;
+
+            case SyntaxKind.Constructor:
+                // A Constructor is ES6 syntax.
+                excludeFlags = TransformFlags.ConstructorExcludes;
+                transformFlags |= TransformFlags.AssertES6;
+
+                // An overload constructor is TypeScript syntax.
+                if (!(<ConstructorDeclaration>node).body) {
+                    transformFlags |= TransformFlags.AssertTypeScript;
+                }
+
+                break;
+
+            case SyntaxKind.PropertyDeclaration:
+                // A PropertyDeclaration is TypeScript syntax.
+                transformFlags |= TransformFlags.AssertTypeScript;
+
+                // If the PropertyDeclaration has an initializer, we need to inform its ancestor
+                // so that it handle the transformation.
+                if ((<PropertyDeclaration>node).initializer) {
+                    transformFlags |= TransformFlags.ContainsPropertyInitializer;
+                }
+
+                break;
+
+            case SyntaxKind.MethodDeclaration:
+                // A MethodDeclaration is ES6 syntax.
+                excludeFlags = TransformFlags.MethodOrAccessorExcludes;
+                transformFlags |= TransformFlags.AssertES6;
+
+                // A MethodDeclaration is TypeScript syntax if it is either async, abstract, overloaded,
+                // generic, or has both a computed property name and a decorator.
+                if ((<MethodDeclaration>node).body === undefined
+                    || (<MethodDeclaration>node).typeParameters !== undefined
+                    || node.flags & NodeFlags.Async
+                    || node.flags & NodeFlags.Abstract
+                    || (subtreeFlags & TransformFlags.ContainsDecorators
+                        && subtreeFlags & TransformFlags.ContainsComputedPropertyName)) {

```

Comment:
why you care about `ContainsComputedPropertyName` to set the flag with `AssertTypeScript`?